curvlinops/ekfac.py

```diff
@@ -0,0 +1,638 @@
+            shapes as the input.
+        """
+        self._maybe_compute_ekfac()
+
+        KM: List[Tensor | None] = [None] * len(M)
+
+        for mod_name, param_pos in self._mapping.items():
+            # cache the weight shape to ensure correct shapes are returned
+            if "weight" in param_pos:
+                weight_shape = M[param_pos["weight"]].shape
+
+            # Get the EKFAC approximation components for the current module
+            aaT_eigenvectors = self._input_covariances_eigenvectors.get(mod_name)
+            ggT_eigenvectors = self._gradient_covariances_eigenvectors.get(mod_name)
+            corrected_eigenvalues = self._corrected_eigenvalues[mod_name]
+
+            # bias and weights are treated jointly
+            if (
+                not self._separate_weight_and_bias
+                and "weight" in param_pos.keys()
+                and "bias" in param_pos.keys()
+            ):
+                w_pos, b_pos = param_pos["weight"], param_pos["bias"]
+                # v denotes the free dimension for treating multiple vectors in parallel
+                M_w = rearrange(M[w_pos], "c_out ... v -> c_out (...) v")
+                M_joint = cat([M_w, M[b_pos].unsqueeze(-2)], dim=-2)
+                M_joint = self._left_and_right_multiply(
+                    M_joint, aaT_eigenvectors, ggT_eigenvectors, corrected_eigenvalues
+                )
+                w_cols = M_w.shape[1]
+                KM[w_pos], KM[b_pos] = M_joint.split([w_cols, 1], dim=-2)
+                KM[b_pos].squeeze_(1)
+            else:
+                self._separate_left_and_right_multiply(
+                    KM,
+                    M,
+                    param_pos,
+                    aaT_eigenvectors,
+                    ggT_eigenvectors,
+                    corrected_eigenvalues,
+                )
+
+            # restore original shapes
+            if "weight" in param_pos:
+                KM[param_pos["weight"]] = KM[param_pos["weight"]].view(weight_shape)
+
+        return KM
+
+    def _compute_eigenvectors(self):
+        """Compute the eigenvectors of the KFAC approximation."""
+        if not (self._input_covariances or self._gradient_covariances):
+            self.compute_kronecker_factors()
+
+        for mod_name in self._mapping.keys():
+            # Free up memory by deleting the Kronecker factors
+            aaT = self._input_covariances.pop(mod_name, None)
+            ggT = self._gradient_covariances.pop(mod_name, None)
+
+            # Compute eigendecomposition of the Kronecker factors and cache eigenvectors
+            if aaT is not None:
+                _, aaT_eigvecs = eigh(aaT)
+                self._input_covariances_eigenvectors[mod_name] = aaT_eigvecs
+                del aaT
+            if ggT is not None:
+                _, ggT_eigvecs = eigh(ggT)
+                self._gradient_covariances_eigenvectors[mod_name] = ggT_eigvecs
+                del ggT
```

Comment:
Refactor into a loop:
```python
for source, destination in zip(
    (self._input_covariances, self._gradient_covariances), 
    (self._input_covariances_eigenvectors, self._gradient_covariances_eigenvectors)
):
    factor = source.pop(mod_name, None)
    if factor is not None:
        destination[mod_name] = eigh(factor)[1]
        # I think we don't need the `del` statement any more because `factor` will be overwritten
        # in the next iteration of the for loop and this will decrease the ref count of the previous 
        # factor to zero, which should trigger garbage collection
```